### PR TITLE
Move THCCachingAllocator to c10_cuda.

### PR DESCRIPTION
**Pull request by ezyang**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #16117 Remove unnecessary includes and headers from THCCachingAllocator, move to at::cuda:: namespace&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13717836/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#16119 Move THCCachingAllocator to c10_cuda.**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13718768/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #16226 Delete duplicate copy of THCCachingAllocator.&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13762540/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #16332 Change default Caffe2 CUDA memory pool to "thc".&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13805475/)

Some renaming and renamespacing also took place. I was originally planning not to do anything, but it turns out that it was easier to make HIPify work by using a namespace CUDACachingAllocator:: rather than THCCachingAllocator_, since :: is a word boundary but _ is not.

Differential Revision: [D13718768](https://our.internmc.facebook.com/intern/diff/D13718768/)

### CODE REVIEW

File: aten/src/ATen/hip/impl/HIPAllocatorMasqueradingAsCUDA.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+#include <c10/core/DeviceType.h>
+
+// Use of c10::hip namespace here makes hipification easier, because
+// I don't have to also fix namespaces.  Sorry!
+namespace c10 { namespace hip {
+
+// Takes a valid HIPAllocator (of any sort) and turns it into
+// an allocator pretending to be a CUDA allocator.  See
+// Note [Masquerading as CUDA]
+class HIPAllocatorMasqueradingAsCUDA : public Allocator {
```
Comment (smessmer) [inline aten/src/ATen/hip/impl/HIPAllocatorMasqueradingAsCUDA.h]: nit: `final`

File: aten/src/ATen/hip/impl/HIPAllocatorMasqueradingAsCUDA.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  Allocator* allocator_;
+public:
+  HIPAllocatorMasqueradingAsCUDA(Allocator* allocator)
```
Comment (smessmer) [inline aten/src/ATen/hip/impl/HIPAllocatorMasqueradingAsCUDA.h]: nit: `explicit`

File: c10/core/Allocator.h [hunk 3]
```diff
@@ -59,6 +59,13 @@ class DataPtr {
   Device device() const {
     return device_;
   }
+  // Unsafely mutates the device on a DataPtr.  Under normal use,
+  // you should never actually need to call this function.
+  // We need this for the implementation of the hack detailed
+  // in Note [Masquerading as CUDA]
+  void unsafe_set_device(Device device) {
```
Comment (smessmer) [inline c10/core/Allocator.h]: Should we `friend` it instead? If you offer a function, people are going to use it, no matter what the comment says ;)

**Review by smessmer — APPROVED** [review]

File: c10/core/Allocator.h (hunk 3 above)
Comment (ezyang) [reply c10/core/Allocator.h]: It's a perfectly legitimate function and can be implemented (less conveniently) in user-space, so I don't think it is worth friend'ing.

**Pull request closed without merging**